tsql-event-store.ddl

```diff
@@ -0,0 +1,218 @@
+END;
+GO
+
+CREATE OR ALTER PROCEDURE append_event
+    @entity       NVARCHAR(255),
+    @entity_key   NVARCHAR(255),
+    @event        NVARCHAR(255),
+    @data         JSON,
+    @append_key   NVARCHAR(255),
+    @previous_id  UNIQUEIDENTIFIER = NULL,
+    @timestamp    DATETIMEOFFSET   = NULL,
+    @event_id     UNIQUEIDENTIFIER OUTPUT
+AS
+BEGIN
+    SET NOCOUNT ON;
+    SET XACT_ABORT ON;
+
+    BEGIN TRY
+        -- set session flag
+        EXEC sys.sp_set_session_context 
+            @key = N'allow_direct_ledger_insert', 
+            @value = 1;
+
+        BEGIN TRAN;
+
+        -- Validate if stream exists already, if previous_id is NULL
+        IF (@previous_id IS NULL)
+        BEGIN
+            IF EXISTS (
+                SELECT 1
+                FROM ledger l
+                WHERE l.entity = @entity
+                  AND l.entity_key = @entity_key
+            )
+            BEGIN
+                THROW 50005, 'previous_id can only be null for first entity event', 1;
+            END;
+        END
+        ELSE
+        BEGIN
+            IF NOT EXISTS (
+                SELECT 1
+                FROM ledger l
+                WHERE l.event_id   = @previous_id
+                  AND l.entity     = @entity
+                  AND l.entity_key = @entity_key
+            )
+            BEGIN
+                THROW 50006, 'previous_id must be in the same entity', 1;
+            END;
+
+            IF EXISTS (
+                SELECT 1
+                FROM ledger prev
+                WHERE prev.event_id = @previous_id
+                  AND EXISTS (
+                      SELECT 1
+                      FROM ledger newer
+                      WHERE newer.entity     = @entity
+                        AND newer.entity_key = @entity_key
+                        AND newer.sequence > prev.sequence
+                  )
+            )
+            BEGIN
+                THROW 50007, 'previous_id must reference the newest event in entity', 1;
+            END;
+        END;
+
+        -- Insert
+        SET @event_id = NEWID();
+
+        INSERT INTO ledger (entity, entity_key, event, data, append_key, previous_id, event_id, timestamp)
+        VALUES (
+            @entity,
+            @entity_key,
+            @event,
+            @data,
+            @append_key,
+            @previous_id,
+            @event_id,
+            ISNULL(@timestamp, SYSDATETIMEOFFSET())
+        );
+
+        COMMIT TRAN;
+
+        -- clear session flag
+        EXEC sys.sp_set_session_context 
+            @key = N'allow_direct_ledger_insert', 
+            @value = NULL;
+    END TRY
+    BEGIN CATCH
+        IF XACT_STATE() <> 0
+            ROLLBACK TRAN;
+
+        -- always clear session flag even on error
+        EXEC sys.sp_set_session_context 
+            @key = N'allow_direct_ledger_insert', 
+            @value = NULL;
+
+        THROW;
+    END CATCH;
+END;
+GO
+
+
+CREATE OR ALTER TRIGGER no_direct_insert_ledger
```

Comment:
This is pretty cool, I should try this approach in the Postgres version (not sure SQLite would be able to do this).